Infra/11750 context documentation (#11875)
google/site-kit-wp: docs/context/js/feature-tours.md

# Feature Tours System

> Site Kit includes a comprehensive feature tours system that guides users through new features and interface changes using interactive tooltips powered by `react-joyride`.

Reviewer: > a comprehensive feature tours system

What does this mean, exactly? 🤔

I don't think our feature tour system is "comprehensive", what did you mean by this? Or is this designed to signify importance to the machine learning model?

## Core Architecture

### Feature Tours Structure

Feature tours are defined as objects with specific properties and stored in `/assets/js/feature-tours/` directory:

```javascript
const myFeatureTour = {
    slug: 'myFeatureTour',                    // Unique identifier
    version: '1.25.0',                       // Site Kit version when tour was added
    contexts: [                              // Where tour can be shown
        VIEW_CONTEXT_MAIN_DASHBOARD,
        VIEW_CONTEXT_ENTITY_DASHBOARD
    ],

>     gaEventCategory: 'main_dashboard_tour',   // Analytics category (string or function)

Author: Not really necessary because gemini will try to review the current state of the codebase anyway and will read the signature that we use for these functions.

    },
    callback: (data, registry) => {          // Optional callback for tour events
        // Handle tour interactions
    }
};
```

### Feature Tours Registration

Tours are imported and exported from the main tours index:

```javascript
// /assets/js/feature-tours/index.js
import sharedKeyMetrics from './shared-key-metrics';
import myNewTour from './my-new-tour';

export default [
    sharedKeyMetrics,
    myNewTour,
    // Additional tours...
];
```

## State Management

### Core User Datastore Integration

Feature tours are managed through the `CORE_USER` datastore in `/assets/js/googlesitekit/datastore/user/feature-tours.js`:

#### Tour State Properties
```javascript
const initialState = {
    lastDismissedAt: undefined,              // Timestamp of last tour dismissal
    dismissedTourSlugs: undefined,           // Array of dismissed tour slugs
    tours: featureTours,                     // All available tours
    currentTour: undefined,                  // Currently active tour
    shownTour: undefined,                    // Tour shown in current page view
};
```

#### Key Actions
```javascript
import { useDispatch } from 'googlesitekit-data';

> import { CORE_USER } from '../googlesitekit/datastore/user/constants';

Reviewer: Shouldn't we avoid relative imports if we're training a model on these examples?

    select(CORE_USER).isTourDismissed('myTourSlug')
);

// Check if tours are on cooldown (2 hours after last dismissal)
const areToursOnCooldown = useSelect((select) =>
    select(CORE_USER).areFeatureToursOnCooldown()
);

// Get all available tours
const allTours = useSelect((select) =>
    select(CORE_USER).getAllFeatureTours()
);

// Get tour shown in current page view
const shownTour = useSelect((select) =>
    select(CORE_USER).getShownTour()
);
```

### Tour Qualification Logic

Tours must pass several requirements to be shown:

1. **View Context Match**: Tour contexts must include current view context
2. **Version Check**: Tour version must be newer than user's initial Site Kit version
3. **Dismissal Check**: Tour must not have been previously dismissed
4. **Custom Requirements**: Optional `checkRequirements` function must return true
5. **Cooldown Check**: No tours dismissed within last 2 hours

```javascript
// Example tour qualification checking
const tourQualifies = await checkTourRequirements({
    tour,
    viewContext: VIEW_CONTEXT_MAIN_DASHBOARD
});

if (tourQualifies) {
    triggerTour(tour);
}
```

## Tour Components

### FeatureTours Component

> Main component that manages tour rendering and lifecycle:

Reviewer: ```suggestion
This component manages tour rendering and lifecycle—it operates as a "side-effect component". It needs to be rendered in the component tree for feature tours to appear.
```

    PREV: 'feature_tooltip_return',        // When user clicks back
    DISMISS: 'feature_tooltip_dismiss',    // When user dismisses tour
    COMPLETE: 'feature_tooltip_complete',  // When tour is completed
};

// Events are automatically tracked when:
trackEvent(gaEventCategory, GA_ACTIONS.VIEW, stepNumber);
trackEvent(gaEventCategory, GA_ACTIONS.COMPLETE, stepNumber);
```

### TourTooltip Component

Custom tooltip component for individual tour steps:

```javascript
import TourTooltip from '../components/TourTooltip';

// Used internally by TourTooltips
function MyCustomTooltip(props) {
    const {
        step,           // Step configuration
        index,          // Current step index
        size,           // Total number of steps
        backProps,      // Back button props
        primaryProps,   // Primary button props
        closeProps,     // Close button props
        tooltipProps    // Tooltip wrapper props
    } = props;
    
    return (
        <div className="tour-tooltip" {...tooltipProps}>
            <h2>{step.title}</h2>
            <div>{step.content}</div>
            {/* Navigation buttons */}
        </div>
    );
}
```

### JoyrideTooltip Component

> Alternative tooltip component for simple, single-step tours:

Reviewer: ```suggestion
Alternative tooltip component for single-step tours:
```

```javascript
import JoyrideTooltip from '../components/JoyrideTooltip';

function SimpleTooltip() {
    return (
        <JoyrideTooltip
            title="Quick Tip"
            content="This is a simple tooltip"
            target=".my-element"
            dismissLabel="Got it"
            placement="bottom"
            disableOverlay={true}
        />
    );
}
```

## Tour Implementation Patterns

### Basic Tour Definition

```javascript

> // /assets/js/feature-tours/my-new-feature.js

Reviewer: What does this line mean/signify?

            />
        );
        
        expect(getByText('Test Step')).toBeInTheDocument();
        expect(getByText('Test content')).toBeInTheDocument();
    });
    
    it('should track events on interaction', () => {
        const mockTrackEvent = jest.fn();
        global.trackEvent = mockTrackEvent;
        
        const { getByText } = render(
            <TourTooltips
                tourID="testTour"
                steps={mockSteps}
                gaEventCategory="test_category"
            />
        );
        
        fireEvent.click(getByText('Next'));
        
        expect(mockTrackEvent).toHaveBeenCalledWith(
            'test_category',
            'feature_tooltip_advance',
            1
        );
    });
});
```

## Best Practices

### Tour Development
1. **Keep tours concise** - Maximum 3-4 steps for better user experience
2. **Use clear, actionable titles** - Tell users what they'll learn
3. **Provide valuable content** - Focus on benefits, not just features
4. **Test target selectors** - Ensure elements exist before tour shows
5. **Consider mobile experience** - Test tours on different screen sizes

### Requirements and Targeting

> 1. **Be specific with requirements** - Only show tours when truly relevant

Reviewer: What does this mean? What do we expect a machine learning algorithm to do with this suggestion?

This is advice I would give to a person, but I don't think a model has the proper context to understand/decide what is "truly relevant", and especially not if we aren't expanding on it more than that.

We should remove this unless it can be expanded upon.

2. **Use semantic selectors** - Prefer class names over IDs for targeting
3. **Handle dynamic content** - Account for async loading and state changes
4. **Respect user context** - Show appropriate tours for user's role/permissions
5. **Version appropriately** - Set tour versions to match feature releases

### Analytics and Tracking

> 1. **Use consistent naming** - Follow established patterns for event categories

Reviewer: Where can the machine learning model learn about these patterns? What are they?

2. **Track key interactions** - View, advance, dismiss, and complete events

> 3. **Monitor tour performance** - Analyze completion rates and drop-off points

Author: Create a suggestion to remove it and i'll accept it.

4. **Custom tracking for important tours** - Add specific metrics for critical features

> 5. **A/B test tour content** - Experiment with different messaging and flows

Reviewer: Do we do this anywhere? I don't think we run A/B tests. Was this generated or something? 😅

### User Experience
1. **Respect dismissal** - Honor user preferences and cooldown periods
2. **Progressive disclosure** - Introduce features gradually, not all at once
3. **Contextual timing** - Show tours when users are most likely to benefit
4. **Accessible design** - Ensure tours work with screen readers and keyboard navigation

> 5. **Performance consideration** - Don't impact core functionality with tour overhead

Reviewer: These seem like generic recommendations an existing large-language model would write about implementing feature tours and don't seems especially relevant to our recommendations or our tours.

		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'test_category',
			'feature_tooltip_advance',
			1
		);
	} );
} );
```

## Best Practices

### Tour Development

1. **Keep tours concise** - Maximum 3-4 steps for better user experience
2. **Use clear, actionable titles** - Tell users what they'll learn
3. **Provide valuable content** - Focus on benefits, not just features
4. **Test target selectors** - Ensure elements exist before tour shows
5. **Consider mobile experience** - Test tours on different screen sizes

### Requirements and Targeting

1. **Be specific with requirements** - Only show tours when truly relevant
2. **Use semantic selectors** - Prefer class names over IDs for targeting
3. **Handle dynamic content** - Account for async loading and state changes
4. **Respect user context** - Show appropriate tours for user's role/permissions
5. **Version appropriately** - Set tour versions to match feature releases

### Analytics and Tracking

1. **Use consistent naming** - Follow established patterns for event categories
2. **Track key interactions** - View, advance, dismiss, and complete events
3. **Monitor tour performance** - Analyze completion rates and drop-off points

> 4. **Custom tracking for important tours** - Add specific metrics for critical features

Reviewer: ```suggestion
1. **Use consistent naming** - Follow established patterns for event categories
2. **Track key interactions** - View, advance, dismiss, and complete events
```
